source/slang/slang-ir-specialize.cpp

```diff
@@ -770,13 +770,99 @@ struct SpecializationContext
         }
     }
 
+    // Finds any `IRTargetDecoration` from `inst`. Recursively chasing `specialize` chains.
+    IRTargetIntrinsicDecoration* findTargetIntrinsicDecorationRec(IRInst* inst)
+    {
+        while (auto specialize = as<IRSpecialize>(inst))
+        {
+            inst = specialize->getBase();
+        }
+        while (auto genericInst = as<IRGeneric>(inst))
+        {
+            inst = findGenericReturnVal(genericInst);
+        }
+        if (auto decor = inst->findDecoration<IRTargetIntrinsicDecoration>())
+            return decor;
+        return nullptr;
+    }
+
+    // Returns true if the call inst represents a call to
+    // StructuredBuffer::operator[]/Load/Consume methods.
+    bool isBufferLoadCall(IRCall* inst)
+    {
```

Comment:
We have a growing problem in the compiler that we either need to represent an operation as its own `IROp` value, in which case we can easily detect it and special-case it through optimizations/transformations, *or* we need to represent it as a built-in function that can use the existing `call` opcode and not need to be special-cased in emit logic for each target.

It seems like we need a better long-term solution here than looking for things like target-intrinsic decorations. After all, shouldn't the target-intrinsic decorations be target-specific? How can we be sure we will find the target-intrinsic decoration that has `.Load` as its value (the HLSL one), and not one for another target?

I don't think this PR needs to introduce the end-all-be-all solution to this problem, but it might be good to take a stand on the issue.

Maybe rather than use the target-intrinsic decoration, we should introduce a new decoration (with supporting syntax) for operations that have "known semantics" (something like `[knownSemantics(...)]` and the operand of the `knownSemantic` attribute can be an integer value that names a case from an `enum KnownSemantics { ... }`. That would let us identify when a function call is invoking an operation with some known semantics to the compiler, and allow passes to take those semantics into account.